### ISSUE TITLE
Handling Missing Modalities

### ISSUE THREAD

**Issue by mkh1989** [Lucas-rbnt/bmmae#2]
Hi,

Thanks for providing the code.

I am trying to use this repo for pretraining a model with 8 modalities. I am using the pretrain_bmmae.py script, and I get an error when a patient does not have all available modalities in the dataset. The problem originates from the following snippet:

`train_files = [
        {
            modality: os.path.join(args.data_dir, "archive", patient, f"{patient}_{modality}.nii.gz")
            for modality in args.modalities
        }
        for patient in train_patients
    ]
`

` val_files = [
        {
            modality: os.path.join(args.data_dir, "val", patient, f"{patient}_{modality}.nii.gz")
            for modality in args.modalities
        }
        for patient in val_patients
 ]
    `

where, for a patient, paths for all modalities are generated. This problem can be seen in the training loop and augmentations as well.

The other problem is related to the masking strategy, where all tokens belonging to a specific modality can be masked, leading to NaN values in the loss function. In line 51 in the loops.py, mask[modality].sum() equals zero, leading to NaN generation in the following line:
`temp_loss= (temp_loss * mask[modality]).sum() / [mask[modality].sum()`

Could you please help me to solve these issues?

Thank you in advance.
Aquila

**Comment by Lucas-rbnt** [comment]
Hi, 

Thank you very much, I'm glad you're using our work.

The code provided is used to support the paper where the handling of missing modalities is done during fine-tuning so it's not possible directly but by design, the architecture handles it so you just need to modify the pre-processing a bit.

In the case of missing modalities, where does the error occur? In fact, this may be due to the batch, which clearly poses a problem, given that a number of tokens will be randomly drawn according to the modalities present and their proportions, and the same number of tokens is expected as input in order to be able to batch properly.

Plus, if a modality is completely masked, then its mask sums up to the number of tokens (1 is for masked, 0 for keep). Have you defined the mask by hand to manage cases where a modality is missing?

Just out of curiosity, what data/modalities do you use?

Have a nice day!

**Comment by mkh1989** [comment]
Hi Lucas,

**1- In the case of missing modalities, where does the error occur?**

The error seems to occur in any part of the repository where there's a loop over the default modalities defined in args.modalities. Below are some example links pointing to the relevant sections:

https://github.com/Lucas-rbnt/BM-MAE/blob/6f08e746df506b983602c39ddabfa0d972d743d0/pretrain_bmmae.py#L73C1-L91C68

https://github.com/Lucas-rbnt/BM-MAE/blob/6f08e746df506b983602c39ddabfa0d972d743d0/bmmae/augmentations.py#L45C1-L83C29

https://github.com/Lucas-rbnt/BM-MAE/blob/6f08e746df506b983602c39ddabfa0d972d743d0/bmmae/loops.py#L38

https://github.com/Lucas-rbnt/BM-MAE/blob/6f08e746df506b983602c39ddabfa0d972d743d0/bmmae/loops.py#L42C13-L47C69


**2- Have you defined the mask by hand to manage cases where a modality is missing?**

No, have not changed the code. The following link point to where I get NaN values when a modality is completely masked.
https://github.com/Lucas-rbnt/BM-MAE/blob/6f08e746df506b983602c39ddabfa0d972d743d0/bmmae/loops.py#L51

**3- Just out of curiosity, what data/modalities do you use?**

I am working on a few datasets including  the following modalities:
["t1", "t2s", "t2", "flair", "swi", "dwi", "pd"]


Best,

Aquila

**Comment by Lucas-rbnt** [comment]
Hi,

Can you please show me the error you get ? My point is, you should have to change the masking strategy so you can handle batch of data in case of missing modality during pre-training (adjust the number of visible tokens) but i want to know if there is some small errors that can be easily corrected before that point.

Ok then are you sure the error occurs at this line ? On my side it works nice with mask full of 1, can you please show me the mask you get as well as the `temp_loss` before this particular line.

Ok then such a nice project!

Best,
Lucas
